File: bindings/dotnet/src/Aster/Interceptors/MetricsInterceptor.cs
using System.Collections.Concurrent;

namespace Aster.Interceptors;

/// <summary>
/// Collects RED metrics (Rate, Errors, Duration) for each RPC call.
/// Falls back to simple in-memory counters (no external dependency required).
/// </summary>
public class MetricsInterceptor : IInterceptor
{
    private int _started;
    private int _succeeded;
    private int _failed;

    /// <summary>Per-call start times keyed by a call identifier.</summary>
    private readonly ConcurrentDictionary<string, long> _callStarts = new();

    /// <summary>Total RPC calls started.</summary>
    public int Started => _started;

    /// <summary>Total RPC calls that succeeded.</summary>
    public int Succeeded => _succeeded;

    /// <summary>Total RPC calls that failed.</summary>
    public int Failed => _failed;

    public object OnRequest(CallContext ctx, object request)
    {
        Interlocked.Increment(ref _started);

        var callKey = $"{ctx.Service}.{ctx.Method}.{ctx.CallId}";
        _callStarts[callKey] = Environment.TickCount64;

        // Store call key on metadata for retrieval in response/error
        ctx.Metadata["_metrics_call_key"] = callKey;

        return request;
    }

    public object OnResponse(CallContext ctx, object response)
    {
        Interlocked.Increment(ref _succeeded);
        FinishTiming(ctx);
        return response;
    }

    public RpcError? OnError(CallContext ctx, RpcError error)
    {
        Interlocked.Increment(ref _failed);
        FinishTiming(ctx);
        return error;
    }

    /// <summary>
    /// Return a snapshot of in-memory counters.
    /// </summary>
    public Dictionary<string, int> Snapshot()
    {
        int started = _started;
        int succeeded = _succeeded;
        int failed = _failed;
        return new Dictionary<string, int>
        {
            ["started"] = started,
            ["succeeded"] = succeeded,
            ["failed"] = failed,
            ["in_flight"] = started - succeeded - failed,
        };
    }

    private void FinishTiming(CallContext ctx)
    {
        if (ctx.Metadata.TryGetValue("_metrics_call_key", out var callKey))
        {
            _callStarts.TryRemove(callKey, out _);
        }
    }
}

File: bindings/java/aster-runtime/src/main/java/site/aster/interceptors/MetricsInterceptor.java
package site.aster.interceptors;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Collects RED metrics (Rate, Errors, Duration) for RPC calls.
 *
 * <p>Provides in-memory counters via {@link #snapshot()}. OpenTelemetry integration is left as a
 * future extension point -- this implementation covers the in-memory counters needed for
 * diagnostics and testing.
 */
public final class MetricsInterceptor implements Interceptor {

  private final AtomicLong started = new AtomicLong();
  private final AtomicLong succeeded = new AtomicLong();
  private final AtomicLong failed = new AtomicLong();

  /** Tracks start times per call key for duration calculation. */
  private final ConcurrentHashMap<String, Long> callStarts = new ConcurrentHashMap<>();

  @Override
  public Object onRequest(CallContext ctx, Object request) {
    started.incrementAndGet();

    // Start timing using a composite key
    String callKey = ctx.service() + "." + ctx.method() + "." + System.identityHashCode(request);
    callStarts.put(callKey, System.nanoTime());

    // Store the key on metadata so on_response/on_error can find it
    ctx.metadata().put("_metrics_call_key", callKey);

    return request;
  }

  @Override
  public Object onResponse(CallContext ctx, Object response) {
    succeeded.incrementAndGet();
    finishTiming(ctx);
    return response;
  }

  @Override
  public RpcError onError(CallContext ctx, RpcError error) {
    failed.incrementAndGet();
    finishTiming(ctx);
    return error;
  }

  private void finishTiming(CallContext ctx) {
    String callKey = ctx.metadata().remove("_metrics_call_key");
    if (callKey != null) {
      callStarts.remove(callKey);
    }
  }

  /** Returns a snapshot of the in-memory counters. */
  public Map<String, Long> snapshot() {
    long s = started.get();
    long ok = succeeded.get();
    long err = failed.get();
    return Map.of(
        "started", s,
        "succeeded", ok,
        "failed", err,
        "in_flight", s - ok - err);
  }
}
